host/sgx/exception_sim.c

```diff
@@ -0,0 +1,214 @@
+        (oe_sgx_td_t*)((uint8_t*)tcs + OE_TD_FROM_TCS_BYTE_OFFSET);
+    uint64_t ssa_frame_size = td->base.__ssa_frame_size;
+    if (ssa_frame_size == 0)
+    {
+        ssa_frame_size = OE_DEFAULT_SSA_FRAME_SIZE;
+    }
+
+    uint64_t ssa_base_address = (uint64_t)tcs + OE_SSA_FROM_TCS_BYTE_OFFSET;
+
+    // cssa always points to the unfilled ssa.
+    return (
+        sgx_ssa_gpr_t*)(ssa_base_address + cssa * ssa_frame_size * OE_PAGE_SIZE - OE_SGX_GPR_BYTE_SIZE);
+}
+
+static void _update_ssa_from_context(ucontext_t* context)
+{
+    sgx_tcs_t* tcs = (sgx_tcs_t*)(context->uc_mcontext.gregs[REG_RBX]);
+    sgx_ssa_gpr_t* ssa_gpr = _get_ssa_gpr(tcs);
+
+    // Update gpr.
+    ssa_gpr->rax = (uint64_t)(context->uc_mcontext.gregs[REG_RAX]);
+    ssa_gpr->rbx = (uint64_t)(context->uc_mcontext.gregs[REG_RBX]);
+    ssa_gpr->rcx = (uint64_t)(context->uc_mcontext.gregs[REG_RCX]);
+    ssa_gpr->rdx = (uint64_t)(context->uc_mcontext.gregs[REG_RDX]);
+    ssa_gpr->rsp = (uint64_t)(context->uc_mcontext.gregs[REG_RSP]);
+    ssa_gpr->rbp = (uint64_t)(context->uc_mcontext.gregs[REG_RBP]);
+    ssa_gpr->rsi = (uint64_t)(context->uc_mcontext.gregs[REG_RSI]);
+    ssa_gpr->rdi = (uint64_t)(context->uc_mcontext.gregs[REG_RDI]);
+    ssa_gpr->r8 = (uint64_t)(context->uc_mcontext.gregs[REG_R8]);
+    ssa_gpr->r9 = (uint64_t)(context->uc_mcontext.gregs[REG_R9]);
+    ssa_gpr->r10 = (uint64_t)(context->uc_mcontext.gregs[REG_R10]);
+    ssa_gpr->r11 = (uint64_t)(context->uc_mcontext.gregs[REG_R11]);
+    ssa_gpr->r12 = (uint64_t)(context->uc_mcontext.gregs[REG_R12]);
+    ssa_gpr->r13 = (uint64_t)(context->uc_mcontext.gregs[REG_R13]);
+    ssa_gpr->r14 = (uint64_t)(context->uc_mcontext.gregs[REG_R14]);
+    ssa_gpr->r15 = (uint64_t)(context->uc_mcontext.gregs[REG_R15]);
+    ssa_gpr->rip = (uint64_t)(context->uc_mcontext.gregs[REG_RIP]);
+    ssa_gpr->rflags = (uint64_t)(context->uc_mcontext.gregs[REG_EFL]);
+
+    // This flag is checked by virtual dispacher.
+    ssa_gpr->exit_info.as_fields.valid = true;
+}
+
+static void _update_sgx_vector(ucontext_t* context, int sig_num)
+{
+    // Hardcode here must match g_vector_to_exception_code_mapping[] in
+    // enclave/core/sgx/exception.c
+    uint32_t sgx_vector = OE_EXCEPTION_UNKNOWN;
+    switch (sig_num)
+    {
+        case SIGFPE: // OE_EXCEPTION_DIVIDE_BY_ZERO
+            sgx_vector = 0;
+            break;
+        case SIGTRAP: // OE_EXCEPTION_BREAKPOIN
+            sgx_vector = 3;
+            break;
+        case SIGILL: // OE_EXCEPTION_ILLEGAL_INSTRUCTION
+            sgx_vector = 6;
+            break;
+        case SIGSEGV: // OE_EXCEPTION_ACCESS_VIOLATION
+            sgx_vector = 13;
+            break;
+        case SIGBUS: // OE_EXCEPTION_MISALIGNMENT
+            sgx_vector = 17;
+            break;
+    }
+
+    sgx_tcs_t* tcs = (sgx_tcs_t*)(context->uc_mcontext.gregs[REG_RBX]);
+    sgx_ssa_gpr_t* ssa_gpr = _get_ssa_gpr(tcs);
+    ssa_gpr->exit_info.as_fields.vector = sgx_vector;
```

Comment:
I though this need to changed; i.e., cast the `sgx_vector` to char/unsigned char.

Reply:
Fixed, thanks!